src/listen/ldap_sync/proto_ldap_sync_ldap.c

```diff
@@ -0,0 +1,1325 @@
+	ldap_msgfree(result);
+
+	/*
+	 *	If the server does not support any of the relevant controls, we just
+	 *	tidy up - no point in signalling to reconnect.
+	 */
+	if (ldap_conn->directory->sync_type == FR_LDAP_SYNC_NONE) {
+		ERROR("LDAP sync configured for directory which does not support any suitable control");
+		talloc_free(dir_ctx);
+		fr_connection_signal_halt(ldap_conn->conn);
+		return;
+	}
+
+	/*
+	 *	We've done all the preparation work on the LDAP connection, now
+	 *	use normal network event listeners.
+	 */
+	fr_event_fd_delete(el, fd, FR_EVENT_FILTER_IO);
+	fr_network_listen_add(thread->nr, thread->li);
+
+	DEBUG2("Starting sync(s)");
+
+	/*
+	 *	Sync operations start by sending a fake packet to run
+	 *	the load Cookie section in order to retrieve the cookie
+	 */
+	for (i = 0; i < talloc_array_length(inst->sync_config); i++) {
+		fr_dbuff_t		dbuff;
+		fr_dbuff_uctx_talloc_t	tctx;
+		size_t			j, len;
+		sync_config_t		*config = inst->sync_config[i];
+
+		/*
+		 *	Ensure we have access to the thread instance
+		 *	in for the demux callbacks
+		 */
+		inst->sync_config[i]->user_ctx = thread;
+
+		/*
+		 *	Assess the namingContext which applies to this sync
+		 */
+		for (j = 0; j < talloc_array_length(ldap_conn->directory->naming_contexts); j++) {
+			len = strlen(ldap_conn->directory->naming_contexts[j]);
+			if (strlen(config->base_dn) < len) continue;
+
+			if (strncasecmp(&config->base_dn[strlen(config->base_dn)-len],
+					ldap_conn->directory->naming_contexts[j],
+					strlen(ldap_conn->directory->naming_contexts[j])) == 0) {
+				config->root_dn = ldap_conn->directory->naming_contexts[j];
+				break;
+			}
+		}
+
+		/*
+		 *	Set up callbacks based on directory type.
+		 */
+		switch (ldap_conn->directory->sync_type) {
+		case FR_LDAP_SYNC_RFC4533:
+			config->init = rfc4533_sync_init;
+			config->entry = rfc4533_sync_search_entry;
+			config->intermediate = rfc4533_sync_intermediate;
+			config->refresh = rfc4533_sync_refresh_required;
+			break;
+
+		case FR_LDAP_SYNC_PERSISTENT_SEARCH:
+			config->init = persistent_sync_state_init;
+			config->entry = persistent_sync_search_entry;
+			break;
+
+		case FR_LDAP_SYNC_ACTIVE_DIRECTORY:
+			config->init = active_directory_sync_state_init;
+			config->entry = active_directory_sync_search_entry;
+			break;
+
+		default:
+			fr_assert(0);
+		}
+
+		MEM(fr_dbuff_init_talloc(NULL, &dbuff, &tctx, 1024, 4096));
+
+		if (fr_dbuff_in(&dbuff, (uint32_t)FR_LDAP_SYNC_CODE_COOKIE_LOAD) < 0) {
+		dbuff_free:
+			fr_dbuff_free_talloc(&dbuff);
+			goto error;
+		};
+
+		/*
+		 *	For Cookie Load packets, use the sync array index as the packet ID.
+		 */
+		if (fr_dbuff_in(&dbuff, (uint32_t)i) < 0) goto dbuff_free;
+
+		if (serialise_sync_config_values(&dbuff, config) < 0) goto dbuff_free;
+
+		/*
+		 *	If the root DN has been identified, populate the attribute.
+		 */
+		if (config->root_dn) {
```

Comment:
Yeah, all of this should just be adding pairs to a temporary pair list, and then calling the internal encoder

Reply:
You can stick them all in thread local structure if you want to avoid the overhead of all the allocs